nucypher/blockchain/eth/interfaces.py

```diff
@@ -606,39 +615,80 @@ def deploy_contract(self,
+                            contract_version=contract.version)
 
         return contract, receipt  # receipt
 
-    def get_contract_factory(self, contract_name: str) -> Contract:
-        """Retrieve compiled interface data from the cache and return web3 contract"""
+    def find_raw_contract_data(self, contract_name: str, requested_version: str = 'latest') -> Tuple[str, dict]:
         try:
-            interface = self.__raw_contract_cache[contract_name]
+            contract_data = self._raw_contract_cache[contract_name]
         except KeyError:
             raise self.UnknownContract('{} is not a locally compiled contract.'.format(contract_name))
         except TypeError:
-            if self.__raw_contract_cache is NO_COMPILATION_PERFORMED:
+            if self._raw_contract_cache is NO_COMPILATION_PERFORMED:
                 message = "The local contract compiler cache is empty because no compilation was performed."
                 raise self.InterfaceError(message)
             raise
-        else:
-            contract = self.client.w3.eth.contract(abi=interface['abi'],
-                                                   bytecode=interface['bin'],
-                                                   ContractFactoryClass=Contract)
-            return contract
 
-    def _wrap_contract(self, wrapper_contract: Contract, target_contract: Contract) -> Contract:
+        try:
+            return requested_version, contract_data[requested_version]
+        except KeyError:
+            if requested_version != 'latest' and requested_version != 'earliest':
+                raise self.UnknownContract('Version {} of contract {} is not a locally compiled. '
+                                           'Available versions: {}'
+                                           .format(requested_version, contract_name, contract_data.keys()))
+
+        if len(contract_data.keys()) == 1:
+            return next(iter(contract_data.items()))
+
+        # Get the latest or the earliest versions
+        current_version_parsed = (-1, -1, -1)
+        current_version = None
+        current_data = None
+        for version, data in contract_data.items():
+            major, minor, patch = [int(v) for v in version[1:].split(".", 3)]
+            if current_version_parsed[0] == -1 or \
+                requested_version == 'latest' and \
+                (major > current_version_parsed[0] or
+                 major == current_version_parsed[0] and minor > current_version_parsed[1] or
+                 major == current_version_parsed[0] and minor == current_version_parsed[1] and patch > current_version_parsed[2]) or \
+                requested_version == 'earliest' and \
+                (major < current_version_parsed[0] or
+                 major == current_version_parsed[0] and minor < current_version_parsed[1] or
+                 major == current_version_parsed[0] and minor == current_version_parsed[1] and patch < current_version_parsed[2]):
```

Comment:
Wow, a 9-line condition. This is though!
I think this part needs some touches to improve readability. As far as I can tell, you're trying to check lexicographical ordering between tuples, which [Python implements already](https://docs.python.org/3/tutorial/datastructures.html#comparing-sequences-and-other-types). So, e.g., lines 667 to 669 can be changed to `(major, minor, patch) > current_version_parsed`; similarly, lines 671 to 673 can be changed to `(major, minor, patch) < current_version_parsed`

Reply:
@cygnusv That's pretty cool - didn't know you could do that.

Reply:
Thanks, really cool suggestion. 
✔️